godot-ffi/src/binding/mod.rs

```diff
@@ -0,0 +1,328 @@
+
+#[cfg(feature = "experimental-threads")]
+pub use multi_threaded::GdextConfig;
+#[cfg(not(feature = "experimental-threads"))]
+pub use single_threaded::GdextConfig;
+
+pub(crate) type GodotBinding = GodotBindingRaw<GdextConfig>;
+
+// Note, this is `Sync` and `Send` because all its fields are. We have avoided implementing `Sync` and `Send` for `GodotBindingRaw`
+// as that could hide issues if any of the field types are changed to no longer be sync/send, but the manual implementation for
+// `GodotBindingRaw` wouldn't detect that.
+//
+// This struct is generic over `Config`, this is only to make it possible to use two different implementations of `GdextConfig`.
+// As we cannot safely use the same `GdextConfig` in both single and multi-threaded code without requiring extra unsafe in `godot-core`.
+// This is because we'd need to implement `Sync` for `GdextConfig`, since it'd be used in the multi-threaded version. But then you could
+// share references to it between threads, which is not safe to do.
+pub(crate) struct GodotBindingRaw<Config> {
+    interface: GDExtensionInterface,
+    library: ClassLibraryPtr,
+    global_method_table: BuiltinLifecycleTable,
+    class_server_method_table: UnsafeOnceCell<ClassServersMethodTable>,
+    class_scene_method_table: UnsafeOnceCell<ClassSceneMethodTable>,
+    class_editor_method_table: UnsafeOnceCell<ClassEditorMethodTable>,
+    builtin_method_table: UnsafeOnceCell<BuiltinMethodTable>,
+    utility_function_table: UtilityFunctionTable,
+    runtime_metadata: GdextRuntimeMetadata,
+    config: Config,
+}
+
+impl<Config> GodotBindingRaw<Config> {
+    pub fn new(
+        interface: GDExtensionInterface,
+        library: GDExtensionClassLibraryPtr,
+        global_method_table: BuiltinLifecycleTable,
+        utility_function_table: UtilityFunctionTable,
+        runtime_metadata: GdextRuntimeMetadata,
+        config: Config,
+    ) -> Self {
+        Self {
+            interface,
+            library: ClassLibraryPtr(library),
+            global_method_table,
+            class_server_method_table: UnsafeOnceCell::new(),
+            class_scene_method_table: UnsafeOnceCell::new(),
+            class_editor_method_table: UnsafeOnceCell::new(),
+            builtin_method_table: UnsafeOnceCell::new(),
+            utility_function_table,
+            runtime_metadata,
+            config,
+        }
+    }
+}
+
+/// Newtype around `GDExtensionClassLibraryPtr` so we can implement `Sync` and `Send` manually for this.
+struct ClassLibraryPtr(crate::GDExtensionClassLibraryPtr);
+
+// SAFETY: It is safe to have access to the library pointer from any thread, as we ensure any access is thread-safe through other means.
+// Even without "experimental-threads", there is no way without `unsafe` to cause UB by accessing the library from different threads.
+unsafe impl Sync for ClassLibraryPtr {}
+// SAFETY: See `Sync` impl safety doc.
+unsafe impl Send for ClassLibraryPtr {}
```

Comment:
Maybe this could mention that "accessing" here refers to read/write the data on a pure Rust level.

_Calling_ functions in Godot is not automatically thread-safe, but that's beyond the responsibility of `Send`/`Sync` here and needs to be enforced by higher-level APIs.

(I guess that's what you meant with "other means"; I guess it doesn't hurt to be explicit though).